intermediate_source/torch_export_tutorial.py

```diff
@@ -0,0 +1,533 @@
+# The first argument of ``dynamic_dim`` is the tensor variable we wish
+# to specify a dimension constraint for. The second argument specifies
+# the dimension of the first argument the constraint applies to.
+# In the example below, our input
+# ``inp1`` has an unconstrained first dimension, but the size of the second
+# dimension must be in the interval (3, 18].
+
+from torch.export import dynamic_dim
+
+inp1 = torch.randn(10, 10)
+
+def constraints_example1(x):
+    x = x[:, 2:]
+    return torch.relu(x)
+
+constraints1 = [
+    dynamic_dim(inp1, 0),
+    3 < dynamic_dim(inp1, 1),
+    dynamic_dim(inp1, 1) <= 18,
+]
+
+exported_constraints_example1 = export(constraints_example1, (inp1,), constraints=constraints1)
+
+print(exported_constraints_example1(torch.randn(5, 5)))
+
+try:
+    exported_constraints_example1(torch.randn(8, 1))
+except Exception:
+    tb.print_exc()
+
+try:
+    exported_constraints_example1(torch.randn(8, 20))
+except Exception:
+    tb.print_exc()
+
+######################################################################
+# Note that if our example inputs to ``torch.export`` do not satisfy the constraints,
+# then we get an error.
+
+constraints1_bad = [
+    dynamic_dim(inp1, 0),
+    10 < dynamic_dim(inp1, 1),
+    dynamic_dim(inp1, 1) <= 18,
+]
+try:
+    export(constraints_example1, (inp1,), constraints=constraints1_bad)
+except Exception:
+    tb.print_exc()
+
+######################################################################
+# We can also use ``dynamic_dim`` to enforce expected equalities between
+# dimensions, for example, in matrix multiplication:
+
+inp2 = torch.randn(4, 8)
+inp3 = torch.randn(8, 2)
+
+def constraints_example2(x, y):
+    return x @ y
+
+constraints2 = [
+    dynamic_dim(inp2, 0),
+    dynamic_dim(inp2, 1) == dynamic_dim(inp3, 0),
+    dynamic_dim(inp3, 1),
+]
+
+exported_constraints_example2 = export(constraints_example2, (inp2, inp3), constraints=constraints2)
+
+print(exported_constraints_example2(torch.randn(2, 16), torch.randn(16, 4)))
+
+try:
+    exported_constraints_example2(torch.randn(4, 8), torch.randn(4, 2))
+except Exception:
+    tb.print_exc()
+
+######################################################################
+# We can actually use ``torch.export`` to guide us as to which constraints
+# are necessary. We can do this by relaxing all constraints (recall that if we
+# do not provide constraints for a dimension, the default behavior is to constrain
+# to the exact shape value of the example input) and letting ``torch.export``
+# error out.
+
+inp4 = torch.randn(8, 16)
+inp5 = torch.randn(16, 32)
+
+def constraints_example3(x, y):
+    if x.shape[0] <= 16:
+        return x @ y[:, :16]
+    return y
+
+constraints3 = (
+    [dynamic_dim(inp4, i) for i in range(inp4.dim())] +
+    [dynamic_dim(inp5, i) for i in range(inp5.dim())]
+)
+
+try:
+    export(constraints_example3, (inp4, inp5), constraints=constraints3)
+except Exception:
+    tb.print_exc()
+
```

Comment:
It might be good to separate these two blocks -- first block prints the stacktrace, and then second block shows that now with the specify_constraints() function, it works.